Jenkinsfile

```diff
@@ -0,0 +1,181 @@
+pipeline {
+    agent any
+    environment {
+        DOCKER_CREDENTIALS_ID = 'c1269293-12a7-4ae4-a1fe-b048736d5658'
+        SLACK_CHANNEL = '#pipeline-notifications'
+        SLACK_TOKEN_CREDENTIAL_ID = 'slack-token'
+    }
+
+    stages {
+        stage('Checkout') {
+            steps {
+                script {
+                    // Checkout your source code from your GitHub repository
+                    def scmVars = checkout scm: [
+                        $class: 'GitSCM',
+                        branches: [[name: '*/develop']],
+                        doGenerateSubmoduleConfigurations: false,
+                        extensions: [],
+                        userRemoteConfigs: [[url: 'https://github.com/imzainazm/microservices-chat-demo.git']]
+                    ]
+                    currentBuild.description = "Triggered by: ${scmVars.GIT_AUTHOR_NAME}"
+                }
+            }
+        }
+
+        stage('Determine Changes') {
+            steps {
+                script {
+                    // Determine which services have changed
+                    def changedFiles = sh(
+                        script: 'git diff --name-only HEAD~1 HEAD',
+                        returnStdout: true
+                    ).trim().split('\n')
+                    
+                    env.CHANGED_SERVICES = []
+                    if (changedFiles.any { it.startsWith('api-gateway/') }) {
+                        env.CHANGED_SERVICES += 'api-gateway'
+                        echo "Changed Services: ${env.CHANGED_SERVICES}"
+                    }
+                    if (changedFiles.any { it.startsWith('users-service/') }) {
+                        env.CHANGED_SERVICES += 'users-service'
+                    }
+                    if (changedFiles.any { it.startsWith('chat-service/') }) {
+                        env.CHANGED_SERVICES += 'chat-service'
+                    }
+                    if (changedFiles.any { it.startsWith('chat-app/') }) {
+                        env.CHANGED_SERVICES += 'chat-app'
+                    }
+                }
+            }
+        }
+
+        stage('Build Docker Images') {
+            when {
+                expression { env.CHANGED_SERVICES }
+            }
+            parallel {
+                stage('Build API Gateway') {
+                    when {
+                        expression { env.CHANGED_SERVICES.contains('api-gateway') }
+                    }
+                    steps {
+                        script {
+                            sh 'docker build -t imzainazm/api-gateway:latest ./api-gateway'
```

Comment:
Added the commit short hash as image tag, did not add the latest tag as it was making the tag long without having of any use as every new image was tagged with latest.

Reply:
I meant you should tag the image with 2 different tags
- api-gateway:f5sy2j3
- api-gateway:latest

so on next it would be, 
- api-gateway:b2mgsl
- api-gateway:latest

This will maintain a history of all the images pushed and also will always have the latest commit tagged as latest image.

Reply:
Already done.